pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/OperationStats.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.oom;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import net.sourceforge.pmd.lang.java.ast.ASTMethodOrConstructorDeclaration;
+import net.sourceforge.pmd.lang.java.oom.Metrics.OperationMetricKey;
+
+/**
+ * Statistics for an operation. Keeps a map of all memoized metrics results.
+ *
+ * @author Clément Fournier
+ */
+class OperationStats {
+
+    private final String name;
+    private final Map<OperationMetricKey, Double> memo = new HashMap<>();
+
+
+    OperationStats(String name) {
+        this.name = name;
+    }
+
+    String getName() {
+        return name;
+    }
+
+    /**
+     * Computes the value of a metric for an operation.
+     *
+     * @param key   The operation metric for which to find a memoized result.
+     * @param node  The AST node of the operation.
+     * @param force Force the recomputation. If unset, we'll first check for a memoized result.
+     *
+     * @return The result of the computation, or {@code Double.NaN} if it couldn't be performed.
+     */
+    double compute(Metrics.OperationMetricKey key, ASTMethodOrConstructorDeclaration node, boolean force) {
+        Double prev = memo.get(key);
+        if (!force && prev != null) {
+            return prev;
+        }
+
+        OperationMetric metric = key.getCalculator();
+        double val = metric.computeFor(node, Metrics.getTopLevelPackageStats());
+        memo.put(key, val);
+        return val;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        } else if (o instanceof String) {
```

Comment:
I regenerated equals and hashcode, and reimplemented the operation lookup map in ClassStats with a map instead of a set as you suggested, which removes the need for an `equalsName` ;)